File: src/core/device/devicemanager.cpp
/**
 * @file devicemanager.cpp
 * @brief 设备管理器类实现
 */

#include "core/device/devicemanager.h"
#include "core/network/devicediscoverer.h"
#include "common/logger/logger.h"
#include <QSysInfo>
#include <QStandardPaths>
#include <QNetworkInterface>

DeviceListModel::DeviceListModel(QObject* parent)
    : QAbstractListModel(parent)
{
}

int DeviceListModel::rowCount(const QModelIndex& parent) const
{
    if (parent.isValid())
    {
        return 0;
    }
    return m_devices.size();
}

QVariant DeviceListModel::data(const QModelIndex& index, int role) const
{
    if (!index.isValid() || index.row() >= m_devices.size())
    {
        return QVariant();
    }

    const DeviceInfo& device = m_devices.at(index.row());

    switch (role)
    {
    case DeviceIdRole:
        return device.deviceId;
    case DeviceNameRole:
        return device.deviceName;
    case IpAddressRole:
        return device.ipAddress;
    case OsTypeRole:
        return device.osType;
    case StatusRole:
        return static_cast<int>(device.status);
    case ActiveTransfersRole:
        return device.activeTransfers;
    case LastSeenRole:
        return device.lastSeen;
    case Qt::DisplayRole:
        return device.deviceName;
    default:
        return QVariant();
    }
}

QHash<int, QByteArray> DeviceListModel::roleNames() const
{
    QHash<int, QByteArray> roles;
    roles[DeviceIdRole] = "deviceId";
    roles[DeviceNameRole] = "deviceName";
    roles[IpAddressRole] = "ipAddress";
    roles[OsTypeRole] = "osType";
    roles[StatusRole] = "status";
    roles[ActiveTransfersRole] = "activeTransfers";
    roles[LastSeenRole] = "lastSeen";
    return roles;
}

void DeviceListModel::addDevice(const DeviceInfo& device)
{
    if (m_deviceIndexMap.contains(device.deviceId))
    {
        return;
    }

    int index = m_devices.size();
    beginInsertRows(QModelIndex(), index, index);
    m_devices.append(device);
    m_deviceIndexMap.insert(device.deviceId, index);
    endInsertRows();
}

void DeviceListModel::updateDevice(const DeviceInfo& device)
{
    if (!m_deviceIndexMap.contains(device.deviceId))
    {
        addDevice(device);
        return;
    }

    int index = m_deviceIndexMap.value(device.deviceId);
    m_devices[index] = device;
    emit dataChanged(createIndex(index, 0), createIndex(index, 0));
}

void DeviceListModel::removeDevice(const QString& deviceId)
{
    if (!m_deviceIndexMap.contains(deviceId))
    {
        return;
    }

    int index = m_deviceIndexMap.value(deviceId);
    beginRemoveRows(QModelIndex(), index, index);
    m_devices.removeAt(index);
    m_deviceIndexMap.remove(deviceId);
    
    for (auto it = m_deviceIndexMap.begin(); it != m_deviceIndexMap.end(); ++it)
    {
        if (it.value() > index)
        {
            it.value()--;
        }
    }
    endRemoveRows();
}

void DeviceListModel::clearDevices()
{
    beginResetModel();
    m_devices.clear();
    m_deviceIndexMap.clear();
    endResetModel();
}

DeviceInfo DeviceListModel::getDevice(int index) const
{
    if (index >= 0 && index < m_devices.size())
    {
        return m_devices.at(index);
    }
    return DeviceInfo();
}

DeviceInfo DeviceListModel::getDevice(const QString& deviceId) const
{
    if (m_deviceIndexMap.contains(deviceId))
    {
        return m_devices.at(m_deviceIndexMap.value(deviceId));
    }
    return DeviceInfo();
}

QList<DeviceInfo> DeviceListModel::getAllDevices() const
{
    return m_devices;
}

int DeviceListModel::findDeviceIndex(const QString& deviceId) const
{
    return m_deviceIndexMap.value(deviceId, -1);
}

DeviceManager& DeviceManager::instance()
{
    static DeviceManager instance;
    return instance;
}

DeviceManager::DeviceManager()
    : m_deviceListModel(nullptr)
{
    m_deviceListModel = new DeviceListModel(this);
}

DeviceManager::~DeviceManager()
{
}

void DeviceManager::initialize()
{
    DeviceInfoData localData;
    localData.deviceId = DeviceDiscoverer::instance().getLocalDeviceInfo().deviceId;
    localData.deviceName = QSysInfo::machineHostName();
    localData.osType = QSysInfo::prettyProductName();
    localData.userName = QStandardPaths::writableLocation(QStandardPaths::HomeLocation).split("/").last();
    
#ifdef Q_OS_WIN
    localData.userName = QStandardPaths::writableLocation(QStandardPaths::HomeLocation).split("\\").last();
#endif

    QList<QHostAddress> addresses = QNetworkInterface::allAddresses();
    for (const QHostAddress& address : addresses)
    {
        if (address.protocol() == QAbstractSocket::IPv4Protocol && 
            !address.isLoopback())
        {
            localData.ipAddress = address.toString();
            break;
        }
    }

    localData.transferPort = NetworkConstants::TRANSFER_PORT;
    
    m_localDevice = DeviceInfo(localData);
    m_localDevice.status = DeviceStatus::Online;

    connect(&DeviceDiscoverer::instance(), &DeviceDiscoverer::deviceDiscovered,
            this, &DeviceManager::onDeviceDiscovered);
    connect(&DeviceDiscoverer::instance(), &DeviceDiscoverer::deviceOffline,
            this, &DeviceManager::onDeviceOffline);
    connect(&DeviceDiscoverer::instance(), &DeviceDiscoverer::deviceUpdated,
            this, &DeviceManager::onDeviceUpdated);

    Logger::instance().info("设备管理器初始化完成，本地设备: " + m_localDevice.deviceName);
}

DeviceListModel* DeviceManager::getDeviceListModel()
{
    return m_deviceListModel;
}

QList<DeviceInfo> DeviceManager::getOnlineDevices() const
{
    QList<DeviceInfo> devices;
    QList<DeviceInfo> allDevices = m_deviceListModel->getAllDevices();
    for (const DeviceInfo& device : allDevices)
    {
        if (device.status == DeviceStatus::Online || device.status == DeviceStatus::Busy)
        {
            devices.append(device);
        }
    }
    return devices;
}

DeviceInfo DeviceManager::getDevice(const QString& deviceId) const
{
    return m_deviceListModel->getDevice(deviceId);
}

bool DeviceManager::isDeviceOnline(const QString& deviceId) const
{
    DeviceInfo device = getDevice(deviceId);
    return device.status == DeviceStatus::Online || device.status == DeviceStatus::Busy;
}

DeviceInfo DeviceManager::getLocalDevice() const
{
    return m_localDevice;
}

void DeviceManager::updateDeviceTransferStatus(const QString& deviceId, int activeTransfers)
{
    DeviceInfo device = getDevice(deviceId);
    if (device.deviceId.isEmpty())
    {
        return;
    }

    device.activeTransfers = activeTransfers;
    device.status = activeTransfers > 0 ? DeviceStatus::Busy : DeviceStatus::Online;
    device.lastSeen = QDateTime::currentDateTime();
    
    m_deviceListModel->updateDevice(device);
    emit deviceStatusChanged(deviceId, device.status);
}

void DeviceManager::addDeviceTransferBytes(const QString& deviceId, qint64 bytes)
{
    DeviceInfo device = getDevice(deviceId);
    if (device.deviceId.isEmpty())
    {
        return;
    }

    device.totalTransferred += bytes;
    device.lastSeen = QDateTime::currentDateTime();
    
    m_deviceListModel->updateDevice(device);
}

void DeviceManager::onDeviceDiscovered(const DeviceInfoData& deviceInfo)
{
    DeviceInfo device(deviceInfo);
    device.status = DeviceStatus::Online;
    device.lastSeen = QDateTime::currentDateTime();
    
    m_deviceListModel->addDevice(device);
    emit deviceAdded(device);
    
    Logger::instance().info("设备管理器添加设备: " + device.deviceName);
}

void DeviceManager::onDeviceOffline(const QString& deviceId)
{
    DeviceInfo device = getDevice(deviceId);
    if (!device.deviceId.isEmpty())
    {
        device.status = DeviceStatus::Offline;
        m_deviceListModel->updateDevice(device);
        emit deviceStatusChanged(deviceId, DeviceStatus::Offline);
    }
    
    m_deviceListModel->removeDevice(deviceId);
    emit deviceRemoved(deviceId);
    
    Logger::instance().info("设备管理器移除设备: " + deviceId);
}

void DeviceManager::onDeviceUpdated(const DeviceInfoData& deviceInfo)
{
    DeviceInfo existingDevice = getDevice(deviceInfo.deviceId);
    
    DeviceInfo device(deviceInfo);
    device.status = existingDevice.status;
    device.activeTransfers = existingDevice.activeTransfers;
    device.totalTransferred = existingDevice.totalTransferred;
    device.lastSeen = QDateTime::currentDateTime();
    
    m_deviceListModel->updateDevice(device);
    emit deviceUpdated(device);
}

File: src/device/device_manager.rs
#![allow(dead_code)]

//! 设备管理器实现
//!
//! 管理所有发现的设备,提供设备状态跟踪和设备信息查询功能

use crate::logger::Logger;
use crate::network::protocol::DeviceInfoData;
use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

/// 设备状态枚举
#[derive(Debug, Clone, PartialEq)]
pub enum DeviceStatus {
    Offline,
    Online,
    Busy,
    Error,
}

/// 设备信息类
#[derive(Debug, Clone)]
pub struct DeviceInfo {
    pub device_id: String,
    pub device_name: String,
    pub ip_address: String,
    pub transfer_port: u16,
    pub os_type: String,
    pub user_name: String,
    pub status: DeviceStatus,
    pub active_transfers: u32,
    pub total_transferred: u64,
}

impl DeviceInfo {
    /// 从DeviceInfoData创建DeviceInfo
    pub fn from_data(data: DeviceInfoData) -> Self {
        Self {
            device_id: data.device_id,
            device_name: data.device_name,
            ip_address: data.ip_address,
            transfer_port: data.transfer_port,
            os_type: data.os_type,
            user_name: data.user_name,
            status: DeviceStatus::Online,
            active_transfers: 0,
            total_transferred: 0,
        }
    }

    /// 转换为DeviceInfoData
    pub fn to_data(&self) -> DeviceInfoData {
        DeviceInfoData {
            device_id: self.device_id.clone(),
            device_name: self.device_name.clone(),
            ip_address: self.ip_address.clone(),
            transfer_port: self.transfer_port,
            os_type: self.os_type.clone(),
            user_name: self.user_name.clone(),
        }
    }
}

/// 设备管理器
pub struct DeviceManager {
    /// 设备列表
    devices: Arc<RwLock<HashMap<String, DeviceInfo>>>,
    /// 本地设备信息
    local_device: Arc<RwLock<Option<DeviceInfo>>>,
}

/// 全局设备管理实例
static INSTANCE: Lazy<DeviceManager> = Lazy::new(|| {
    DeviceManager {
        devices: Arc::new(RwLock::new(HashMap::new())),
        local_device: Arc::new(RwLock::new(None)),
    }
});

impl DeviceManager {
    /// 获取单例实例
    pub fn instance() -> &'static DeviceManager {
        &INSTANCE
    }

    /// 初始化设备管理器
    pub fn initialize(&self) -> anyhow::Result<()> {
        Logger::info("设备管理器初始化完成");
        Ok(())
    }

    /// 获取所有在线设备
    pub async fn get_online_devices(&self) -> Vec<DeviceInfo> {
        self.devices
            .read()
            .await
            .values()
            .filter(|d| d.status == DeviceStatus::Online || d.status == DeviceStatus::Busy)
            .cloned()
            .collect()
    }

    /// 获取指定设备信息
    pub async fn get_device(&self, device_id: &str) -> Option<DeviceInfo> {
        self.devices.read().await.get(device_id).cloned()
    }

    /// 检查设备是否在线
    pub async fn is_device_online(&self, device_id: &str) -> bool {
        self.devices
            .read()
            .await
            .get(device_id)
            .map(|d| d.status == DeviceStatus::Online || d.status == DeviceStatus::Busy)
            .unwrap_or(false)
    }

    /// 获取本地设备信息
    pub async fn get_local_device(&self) -> Option<DeviceInfo> {
        self.local_device.read().await.clone()
    }

    /// 添加设备
    pub async fn add_device(&self, device_info: DeviceInfo) {
        self.devices
            .write()
            .await
            .insert(device_info.device_id.clone(), device_info.clone());
        Logger::info(&format!("设备管理器添加设备: {}", device_info.device_name));
    }

    /// 移除设备
    pub async fn remove_device(&self, device_id: &str) {
        self.devices.write().await.remove(device_id);
        Logger::info(&format!("设备管理器移除设备: {}", device_id));
    }
}